writeups/picomini-2021-diana/notepad.md

```diff
@@ -0,0 +1,101 @@
+---
+title: picoMini by redpwn 2021 - notepad (web)
+date: 2021-05-12
+slug: /writeups/picomini-redpwn-notepad
+excerpt: Writeup for notepad (web)
+author: Diana Lin
+---
+
+Notepad was a web exploitation challenge from picoMini by redpwn 2021.
+
+# Description
+> This note-taking site seems a bit off.
+>
+> notepad.mars.picoctf.net
+
+Files:
+* [notepad.tar](https://artifacts.picoctf.net/picoMini+by+redpwn/Web+Exploitation/notepad/notepad.tar)
+
+The application is a simple note-taking site. When a new note is created, a file in the static folder is created, and the client is redirected to the file.
+
+# Directory Traversal
+We were provided a source so I checked that out. I first noticed the filters on the note's `content`.
+```py
+if "_" in content or "/" in content:
+    return redirect(url_for("index", error="bad_content"))
+```
+
+The filter on `/` indicates that path traversal should be prevented. However, the created file's `name` looks odd.
+```py
+    name = f"static/{url_fix(content[:128])}-{token_urlsafe(8)}.html"
+    with open(name, "w") as f:
+        f.write(content)
+```
+
+The filename is the first 128 characters of the content and a random token. It's a bit weird that they're normalizing the path with `url_fix(content[:128])`; that's our first vulnerability. Even though `/` is filtered, `\` is not, and `url_fix()` would normalize the backslashes to be `/`. We could use this to traverse the filesystem.
+
+# Arbitrary Template Creation
+Remember that error from the filter on `content`? That comes in handy now. On the index page, the error can be set to any arbitrary value by simply passing in another value to the `error` query parameter:
+```
+https://notepad.mars.picoctf.net/?error=error_name
+```
+
+This value is then used as the filename for a Jinja template in the `errors` subdirectory.
+```html
+{% if error is not none %}
+  <h3>
+    error: {{ error }}
+  </h3>
+  {% include "errors/" + error + ".html" ignore missing %}
+{% endif %}
+```
+
+How is this useful? Well, we can create our own error template by using our path traversal technique to create a file in the `templates/errors` folder. This new template can then be rendered on the index page if we pass it in as a query parameter.
+
+We can test a sample payload now:
+```
+..\\templates\\errors\\
```

Comment:
small nitpick—python string literals need escaped backslashes `\\` but the actual payload only has one backslash